### PR TITLE
docs(appendix): Update docs with new info on configuration

### LINKED ISSUE #3968: Document configuration for Zeebe
**Description**

Expand documentation of the zeebe configuration
* Add example for gateway configuration
* Explain how to specify s in the configuration
* Explain how to use environment variables, e.g. to 
* Explain which configuration options are at least needed

### PR DESCRIPTION
**Pull request by pihme**
## Description

This pull request updates and expands the documentation with respect to the new configuration options.

Some tests were also added to address some edge cases that arose while writing the documentation.

## Related issues

closes #3968

## Pull Request Checklist

- [ X ] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [ X ] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [ X ] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

**Review by npepinpe — CHANGES_REQUESTED** [review]
:+1: 

I like it. In general I only have minor change requests (e.g. typos), let me know if anything is unclear. 

One thing that came to mind is whether or not we should have the legacy configuration in the appendix, or in the operations section under pre 0.23.x configuration?

**Comment by pihme** [comment]
@jwulf Did you have time to take a look at this?

File: docs/src/introduction/install.md [hunk 1]
```diff
@@ -33,7 +33,7 @@ Docker configurations for starting a single Zeebe broker using `docker-compose`,
 You can run Zeebe with Docker:
 
 ```bash
-docker run --name zeebe -p 26500:26500 camunda/zeebe:latest
+docker run --name zeebe -p 26500-26502:2600-26502 camunda/zeebe:latest
```
[the remark this answers is not in the document]
Comment (jwulf) [reply docs/src/introduction/install.md]: I think we need to have the "legacy" configuration in the docs. My assessment is grounded in our statement that 0.20 was production ready - and I know of several users who have in fact gone into production with pre-0.23 brokers. They will need documentation until they are ready to do the work to migrate to a later version in production.

File: docs/src/operations/resource-planning.md [hunk 2]
```diff
@@ -8,64 +8,124 @@ While we cannot tell you exactly what you need - beyond _it depends_ - we can ex
...
+
+leaderPartitionsPerNode = partitionsCount / numberOfNodes 
+followerPartitionsPerNode = partitionsCount * replicationFactor / numberOfNodes 
+
+clusterSize = [number of broker nodes]
+partitionsCount = [number of partitions]
+replicationFactor = [number of replicas per partition]
+reserveForPartialSystemFailure = [factor to account for partial system failure]  
+singleSnapshotSize = [size of a single rocks DB snapshot]                  
+eventLogSize = [event log size for duration of maxSnapshots * snapshotPeriod]
```
[the remark this answers is not in the document]
Comment (jwulf) [reply docs/src/operations/resource-planning.md]: I heard that we are fixing snapshots at 1? Simple, sane defaults are best.

File: docs/src/operations/resource-planning.md [hunk 3]
```diff
@@ -8,64 +8,124 @@ While we cannot tell you exactly what you need - beyond _it depends_ - we can ex
...
 
 A snapshot is a projection of all events that represent the current running state of the workflows running on the partition.  It contains all active data, for example, deployed workflows, active workflow instances, and not yet completed jobs.
 
-When the broker has as many snapshots as configured by the parameter maxSnapshots, it deletes all data on the log which was written before the oldest snapshot.
+When the broker has as many snapshots as configured by the parameter `zeebe.broker.data.maxSnapshots`, it deletes all data on the log which was written before the oldest snapshot.
 
-## RocksDB
+### RocksDB
 
 On the lead broker of a partition, the current running state is kept in memory, and on disk in RocksDB. In our experiments this grows to 2GB under a heavy load of long-running processes. The snapshots that are replicated to followers are snapshots of RocksDB.
```
Comment (jwulf) [inline docs/src/operations/resource-planning.md]: In our experience this grows

File: docs/src/operations/setting-up-a-cluster.md [hunk 4]
```diff
@@ -194,8 +194,8 @@ each node knows exactly, which partitions it has
 to bootstrap and for which it will become the leader at first (this
 could change later, if the node needs to step down for example).
 
-## Keep Alive
+## Keep Alive Intervals
 
 It's possible to specify how often Zeebe clients should send keep alive pings. By default, the official Zeebe clients send keep alive pings every 45 seconds. This interval can be configured through the clients' APIs and through the `ZEEBE_KEEP_ALIVE` environment variable. When configuring the clients with the environment variable, the time interval must be expressed a positive amount of milliseconds (e.g., 45000).
```
Comment (jwulf) [inline docs/src/operations/setting-up-a-cluster.md]: By default, the official Zeebe clients (Java and Go) send keep alive pings

**Review by jwulf — CHANGES_REQUESTED** [review]
Great work! A bunch of minor edits and suggestions - probably the largest keeping the legacy configuration as a reference for existing users.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 566314bacd1a530ad11ed62f64002a00022dae76
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 566314bacd1a530ad11ed62f64002a00022dae76
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: d3a04559f1d562b486923cddbf9c9677a89c110e
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 566314bacd1a530ad11ed62f64002a00022dae76
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: d3a04559f1d562b486923cddbf9c9677a89c110e
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

File: docs/src/operations/configuration.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+*Conventions*
+
+Rename the configuration file to `application.yml` and place it in the following location:
+```shell script
+./config/application.yml
+``` 
+
+*Misc*
+
+Zeebe uses Spring Boot for its configuration parsing. So all other ways to configure a Spring Boot application should also work.
```
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/operations/configuration.md]: done

File: docs/src/operations/resource-planning.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/operations/resource-planning.md]: please look at this section again. I already removed the `maxSnapshot` setting in anticipation of @korthout PR. Still I am unsure about the formulas.

File: docs/src/operations/resource-planning.md [hunk 6]
```diff
@@ -8,64 +8,124 @@ While we cannot tell you exactly what you need - beyond _it depends_ - we can ex
...
 - `segments` - the data of the log split into segments. The log is only appended - until it is truncated by reaching the max snapshot count.
 - `state` - the active state. Deployed workflows, active workflow instances, etc. Completed workflow instances or jobs are removed.
 - `snapshot` - the data of a state at a certain point in time
 
-If you want a recipe to explode your disk space usage, here are a few ways to do it:
+> **Pitfalls** 
+> 
+> If you want to avoid exploding your disk space usage, here are a few pitfalls to avoid:
+> - Do not create a high number of snapshots with a long period between them.
+> - Do not configure an exporter which does not not advance its record position (such as the Debug Exporter)
```
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/operations/resource-planning.md]: done, please look at this section again

File: docs/src/operations/resource-planning.md [hunk 7]
```diff
@@ -8,64 +8,124 @@ While we cannot tell you exactly what you need - beyond _it depends_ - we can ex
...
+ - all the events it contains have been processed by exporters, 
+ - all the events it contains have been replicated to other brokers, 
+ - all the events it contains have been processed, and
+ - the maximum number of snapshots has been reached. 
 
-- A cluster loses its quorum, in which case events are queued but not processed. 
-- An exporter does not advance its read position in the event log.
-- The max number of snapshots has not been written.
+The following conditions inhibit the automatic deletion of event log segments:
+- A cluster loses its quorum. In this case events are queued but not processed. Once a quorum is reestablished, events will be replicated and eventually event log segments will be deleted.
```
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/operations/resource-planning.md]: I don't know how to describe this well. Maybe this is also not so much a topic for ordinary resource planning, but more a topic for how to deal with cascading failures.

File: docs/src/appendix/deprecated-features.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+### Legacy Environment Variables
+With the changes to the configuration, legacy environment variables were also deprecated.
+
+They are still supported during the backwards compatibility window and will be internally mapped to the new environment variables. If legacy environment variables are detected, there will be a warning in the log:
+
+```
+17:20:00.591 [] [main] WARN  io.zeebe.legacy - Found use of legacy system environment variable 'ZEEBE_HOST'. Please use 'ZEEBE_BROKER_NETWORK_HOST' instead.
+17:20:00.602 [] [main] INFO  io.zeebe.legacy - The old environment variable is currently supported as part of our backwards compatibility goals.
+17:20:00.602 [] [main] INFO  io.zeebe.legacy - However, please note that support for the old environment variable is scheduled to be removed for release 0.25.0.
+__  ___  ___  __   ___  ___     __   __   __        ___  __
```
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/appendix/deprecated-features.md]: Don't know. I also see many cases where Zeebe is written in lowercase. Who would know such a thing?

File: docs/src/appendix/environment-variables.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+## Environment Variables for Configuration
+The configuration can be provided as a file and/or through environment variables. Mixing both sources is also possible. In that case environment variables have precedence over the configuration settings in the configuration file. 
+
+The respective environment variables are documented in the configuration file templates:
+* [Broker Configuration Template](broker-config-template.md)
+* [Gateway Configuration Template](gateway-config-template.md)
+
+## Environment Variables for Operators
+The following environment variables are intended for operators:
+  - `ZEEBE_LOG_LEVEL`: Sets the log level of the Zeebe Logger (default: `info`).
```
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/appendix/environment-variables.md]: Would like to keep this. Will work for broker + gateway, afaik

File: docs/src/operations/configuration.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+## Verifying that configuration was applied
+To verify that the configuration was applied, start the application and look at the applications log.
```
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/operations/configuration.md]: Again, would like to keep it this way. It pertains to Broker and Gateway alike.

File: docs/src/operations/configuration.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+If the configuration could be read, the application will log out the effective configuration during startup:
```
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/operations/configuration.md]: same reason

File: docs/src/operations/configuration.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+      "bytes": 4194304
+    },
+    "commandApi": {
+      "defaultPort": 26501,
+      "host": "0.0.0.0",
+      "port": 26501,
+...
+```
+
+In some cases of invalid configuration the application will fail to start with a warning that explains which configuration setting could not be read.
```
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/operations/configuration.md]: same reason

File: docs/src/operations/resource-planning.md [hunk 13]
```diff
@@ -8,64 +8,124 @@ While we cannot tell you exactly what you need - beyond _it depends_ - we can ex
...
+> If you want to avoid exploding your disk space usage, here are a few pitfalls to avoid:
+> - Do not create a high number of snapshots with a long period between them.
+> - Do not configure an exporter which does not not advance its record position (such as the Debug Exporter)
 
-- Create a high number of snapshots with a long period between them.
-- Load an exporter, such as the Debug Exporter, that does not advance its record position.
+### Event Log
 
-## Event Log
+The event log for each partition is segmented. By default, the segment size is 512MB. This can be changed in the [configuration](/appendix/broker-config-template.md) with the setting `zeebe.broker.data.logSegmentSize`.
```
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/operations/resource-planning.md]: Removed the second sentence. Please look at it again whether it is better now

File: docs/src/operations/resource-planning.md (hunk 3 above)
Comment (pihme) [reply docs/src/operations/resource-planning.md]: yeah, that was what I meant to type :-)

**Comment by pihme** [comment]
Went through your comments. Incorporated most. Some questions from my side, and some pushback on some points. Please have a look at it again

File: docs/src/operations/resource-planning.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (npepinpe) [reply docs/src/operations/resource-planning.md]: looks good to me :+1:

File: docs/src/operations/resource-planning.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (npepinpe) [reply docs/src/operations/resource-planning.md]: Let's leave it as is for now then

**Review by npepinpe — APPROVED** [review]
Looks good to me :+1:

File: docs/src/appendix/deprecated-features.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (jwulf) [reply docs/src/appendix/deprecated-features.md]: At the moment it says ZEEBEE with four e's

File: docs/src/operations/configuration.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (jwulf) [reply docs/src/operations/configuration.md]: In that case, maybe:

"To verify that the configuration was applied, start Zeebe and look at the log"

File: docs/src/operations/configuration.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jwulf) [reply docs/src/operations/configuration.md]: In that case, maybe: "Zeebe will log out..."

The reasoning being, that "application" is not a term that we have used to refer to Zeebe previously. Usually we say "the broker".

File: docs/src/operations/configuration.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (jwulf) [reply docs/src/operations/configuration.md]: In some cases of invalid configuration Zeebe will fail to start

**Review by jwulf — CHANGES_REQUESTED** [review]

File: docs/src/appendix/deprecated-features.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/appendix/deprecated-features.md]: There is a dedicated issue for this (which sparked some discussion regarding font) #4111

File: docs/src/operations/configuration.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
+# Configuration
+
```
[the remark this answers is not in the document]
Comment (pihme) [reply docs/src/operations/configuration.md]: I will bring this up with Sebastian or in the team meeting, and create a new issue for this, Here I would like to close it for now, so I can merge the canges

**Comment by pihme** [comment]
Please have a good look at the documentation again. I had some merge conflicts due to the changes by Nico regarding the removal of the `maxSnapshots` setting.

**Review by jwulf — APPROVED** [review]
YOLO

File: docs/src/appendix/deprecated-features.md [hunk 15]
```diff
@@ -0,0 +1,49 @@
+# Deprecated Features
+
+This section lists deprecated features.
+
+## Deprecated in 0.23.0-alpha2
+- TOML configuration - deprecated and removed in 0.23.0-alpha2
+- Legacy environment variables - deprecated in 0.23.0-alpha2, to be removed in 0.25.0
```
Comment (npepinpe) [inline docs/src/appendix/deprecated-features.md]:
```suggestion
- Legacy environment variables - deprecated in 0.23.0-alpha2, to be removed in 0.24.0
```

As far as I know, we support two minor versions, so when we are on 0.24.x, we will be supporting 0.23.x and 0.24.x - so at that point we can already remove legacy configuration support. Correct me if I'm wrong @menski

**Review by npepinpe — APPROVED** [review]
LGTM - only small thing is the 0.25.0 mention, we can already drop support in 0.24.0. Anyway not a blocker, just correct it and merge, no need for another review as far as I can tell :+1:

**Comment by pihme** [comment]
bors retry

**Pull request merged**